internal/nexodus/policy_linux.go

```diff
@@ -0,0 +1,639 @@
+}
+
+func (ax *Nexodus) nfTableExists() (bool, error) {
+	args := []string{"list", "tables"}
+	output, err := runNftCmd(ax.logger, args)
+	if err != nil {
+		return false, err
+	}
+
+	tableFullName := fmt.Sprintf("%s %s", tableFamily, tableName)
+	return strings.Contains(output, tableFullName), nil
+}
+
+// nfCreateTable is used to create the nftables table
+func (ax *Nexodus) nfCreateTable() error {
+	if _, err := runNftCmd(ax.logger, []string{"add", "table", tableFamily, tableName}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// nfCreateChain is used to create the nftables chain in the nf table
+func (ax *Nexodus) nfCreateChain(chainName string) error {
+	if _, err := runNftCmd(ax.logger, []string{"add", "chain", tableFamily, tableName, chainName, "{", "type", "filter", "hook", "input", "priority", "0", ";", "policy", "accept", ";", "}"}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// containsIPv4Range matches the following ipv4 patterns:
+// Cidr notation 100.100.0.0/16
+// Individual address 10.100.0.2
+// Dash-separated range 100.100.0.0-100.100.10.255
+func containsIPv4Range(ipRanges []string) bool {
+	for _, ipRange := range ipRanges {
+		if strings.Contains(ipRange, "-") {
+			// Dash-separated range
+			ips := strings.Split(ipRange, "-")
+			ip1 := net.ParseIP(strings.TrimSpace(ips[0]))
+			ip2 := net.ParseIP(strings.TrimSpace(ips[1]))
+
+			if ip1 != nil && ip1.To4() != nil && ip2 != nil && ip2.To4() != nil {
+				return true
+			}
+		} else if strings.Contains(ipRange, "/") {
+			// CIDR notation
+			_, ipNet, err := net.ParseCIDR(ipRange)
+			if err == nil && ipNet.IP.To4() != nil {
+				return true
+			}
+		} else {
+			ip := net.ParseIP(ipRange)
+			// Individual IP
+			if ip != nil && ip.To4() != nil {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// containsIPv6Range matches the following ipv6 patterns:
+// Cidr notation 200::/64
+// Individual address 200::2
+// Dash-separated range Range 200::1-200::8
+// Dash-separated range 2001:0db8:0000:0000:0000:0000:0000:0000-2001:0db8:ffff:ffff:ffff:ffff:ffff:ffff
+func containsIPv6Range(ipRanges []string) bool {
+	for _, ipRange := range ipRanges {
+		if strings.Contains(ipRange, "-") {
+			// Dash-separated range
+			ips := strings.Split(ipRange, "-")
+			if len(ips) != 2 {
+				return false
+			}
+
+			ip1 := net.ParseIP(strings.TrimSpace(ips[0]))
+			ip2 := net.ParseIP(strings.TrimSpace(ips[1]))
+
+			if ip1 == nil || ip2 == nil || ip1.To16() == nil || ip2.To16() == nil {
+				return false
+			}
+		} else if strings.Contains(ipRange, "/") {
+			// CIDR notation
+			_, _, err := net.ParseCIDR(ipRange)
+			if err != nil {
+				return false
+			}
+		} else {
+			// Individual IP
+			ip := net.ParseIP(ipRange)
+			if ip == nil || ip.To16() == nil {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// runNftCmd is used to execute nft commands
+func runNftCmd(logger *zap.SugaredLogger, cmd []string) (string, error) {
+	nft := exec.Command("nft", cmd...)
```

Comment:
netlink would probably be more efficient ... that's definitely a "file an issue to consider in the future" thing though

Reply:
Yeah, it was discussed in a meetup. I started down that path and at about 5k loc just to add a builder for rules to use the nftable library https://pkg.go.dev/github.com/google/nftables it made more sense to start with the userspace command. netfilter is a big api. Also its very clear what is getting inserted for initial debugging with the CLI which was another factor vs. netlink.

Reply:
I remember that discussion now. Yeah, this is fine for now. I'm also fine not even filing an issue for it until it seems like a problem.